Tasks/DownloadBuildArtifacts/main.ts

```diff
@@ -0,0 +1,151 @@
+import * as providers from 'item-level-downloader/Providers';
+import * as webHandlers from 'item-level-downloader/Providers/Handlers';
+
+tl.setResourcePath(path.join(__dirname, 'task.json'));
+
+async function main(): Promise<void> {
+    var promise = new Promise<void>(async (resolve, reject) => {
+        var buildType: string = tl.getInput("buildType", true);
+        var isCurrentBuild: boolean = buildType.toLowerCase() === 'current';
+        var projectId: string = isCurrentBuild ? tl.getVariable("System.TeamProjectId") : tl.getInput("project", true);
+        var definitionId: string = isCurrentBuild ? '' : tl.getInput("definition", true);
+        var buildId: number = parseInt(isCurrentBuild ? tl.getVariable("Build.BuildId") : tl.getInput("buildId", true));
+        var downloadPath: string = tl.getInput("downloadPath", true);
+        var downloadType: string = tl.getInput("downloadType", true);
+
+        var endpointUrl: string = tl.getVariable("System.TeamFoundationCollectionUri");
+        var accessToken: string = tl.getEndpointAuthorizationParameter('SYSTEMVSSCONNECTION', 'AccessToken', false);
+        var credentialHandler: IRequestHandler = getHandlerFromToken(accessToken);
+        var webApi: WebApi = new WebApi(endpointUrl, credentialHandler);
+        var debugMode: string = tl.getVariable('System.Debug');
+        var isVerbose: boolean = debugMode ? debugMode.toLowerCase() != 'false' : false;
+        var parallelLimit: number = +tl.getVariable("release.artifact.download.parallellimit");
+
+        var templatePath: string = path.join(__dirname, 'vsts.handlebars.txt');
+        var buildApi: IBuildApi = webApi.getBuildApi();
+        var artifacts = [];
+        var itemPattern: string = '**';
+
+        // verfiy that buildId belongs to the definition selected
+        if (definitionId) {
+            var build = await buildApi.getBuild(buildId, projectId).catch((reason) => {
+                reject(reason);
+                return;
+            });
+
+            if (build) {
+                if (build.definition.id !== parseInt(definitionId)) {
+                    reject(tl.loc("BuildIdBuildDefinitionMismatch", buildId, definitionId));
+                    return;
+                }
+            }
+            else {
+                reject(tl.loc("BuildNotFound", buildId));
+                return;
+            }
+        }
+
+        // populate itempattern and artifacts based on downloadType
+        if (downloadType === 'single') {
+            var artifactName = tl.getInput("artifactName");
+            var artifact = await buildApi.getArtifact(buildId, artifactName, projectId).catch((reason) => {
+                reject(reason);
+                return;
+            });
+
+            if (!artifact) {
+                reject(tl.loc("BuildArtifactNotFound", artifactName, buildId));
+                return;
+            }
+
+            artifacts.push(artifact);
+            itemPattern = artifactName + '/**';
```

Comment:
Why is itemPattern always `myartifact/**` in the single case, but in the all case it's either `**` or specified as an input?

Reply:
@scottdallamura itemPattern is a minimatch pattern to select all files under an artifact named drop we will need to provide item pattern as drop/** . ** matches all the files so by default all artifacts will be downloaded. For specific input you can provide any minimatch pattern and only those files will be downloaded, like if one wants to download drop1 and drop2 artifacts he can specify @(drop1|drop2)/**

Reply:
It looks like in the multiple case I can also specify itemPattern="**/*.zip".  In the single case, the itemPattern input is ignored.  Is this by design?  What if I want to get just one file from one artifact?

Reply:
@scottdallamura Yes, this is by design, for the "single case" all the files within that artifact will be downloaded. If he selects "multiple" he is free to give any pattern he likes, depending on pattern we will select the files to be downloaded. If user wants to download only zip files for a single artifact he can select "multiple" and give a pattern like drop/**/*.zip. If he specifies **/*.zip then all the zip files from all the artifacts will be downloaded.

Reply:
This seems like an arbitrary restriction, then.  Why not respect the itemPattern in the single case?

Reply:
@scottdallamura There isn't any restriction here, I think the selection between 'single artifact' and 'multiple artifacts' is confusing. I have updated it to 'single artifact' or 'specific files' which is what it is actually doing. If user wants to download a specific file under a single artifact he can use 'specific files' option.

Reply:
Ok, that makes more sense.  Thanks.